tests/test_middleware.py

```diff
@@ -263,6 +263,28 @@ def test_multi_args_callbacks(settings):
     assert item["non_cb_arg"] is None
 
 
+@pytest.mark.parametrize(
+    "final_kwargs,cb_kwargs,expected",
+    [
+        ({}, {}, {}),
+        ({"x": 1}, {}, {"x": 1}),
+        ({}, {"x": 2}, {"x": 2}),
+        ({"x": 1}, {"x": None}, {"x": 1}),
+        ({"x": 1}, {"x": False}, {"x": 1}),
+        ({"x": 1}, {"x": 2}, {"x": 1}),
+        ({"x": None}, {"x": 2}, {"x": 2}),
+        ({"x": False}, {"x": 2}, {"x": False}),
```

Comment:
This part is tricky since in order for `Retry` to work, the `InjectorMiddleware` needs to replace the `Request.cb_kwargs` values to ensure fresh page objects, responses, and other dependencies.

However, the user loses control when wanting to specifically provide a dependency instead of relying to scrapy-poet's providers.

I'll look into this further.

Reply:
Maybe we should make it so that by default re-injection happens, and later on provide a way to explicitly ask for a cb_kwarg not to be re-injected (maybe some meta key).

Reply:
I was thinking if we should:
- disallow users from overriding dependencies created by providers via `Request.cb_kwargs`
- instead, only allow dependencies from being overridden by something like `Request(..., meta={"SCRAPY_POET_DEPS": {page: MyPage()}})`

This means that:
- https://github.com/scrapinghub/scrapy-poet/issues/110 won't be needed since the `Injector` needs to use the meta key `SCRAPY_POET_DEPS`
- https://github.com/scrapinghub/scrapy-poet/issues/98 would cease to confuse users.

Though I'm not sure if this breaks some important use cases for some users.

Reply:
nvm, I found a way to persist the original user-specified `request.cb_kwargs` across multiple retries. https://github.com/scrapinghub/scrapy-poet/pull/113/commits/0425a70926487e40ffd215c0c76231f8b7380cbd 

I think this solution is much cleaner since:
- it doesn't add a new setting which could make **scrapy-poet** a bit more complex.
- user's have direct control on overriding the callback dependencies via `request.cb_kwargs` itself.

Reply:
As discussed off-GitHub, we're going with the first route discussed previously:

- disallow users from overriding dependencies created by providers via `Request.cb_kwargs`. This avoids the confusion during retries where the values in `cb_kwargs` come from the past `Request` or are actually provided by the user.

This avoids the complexities of doing https://github.com/scrapinghub/scrapy-poet/commit/0425a70926487e40ffd215c0c76231f8b7380cbd which creates complexities and confusion when new middlewares are added before/after the `InjectionMiddleware`.

This approach creates a backward incompatible change of not support user provided dependencies anymore via `Request.cb_kwargs`. https://github.com/scrapinghub/scrapy-poet/pull/113/commits/a1c48c1122ee4e02dfd66bc4b2aef44bb2c145cc